auth changes

msbutler opened the pull request:
contains randomized ROOT setting....?

msbutler committed:
sql/privilege introduce REPLICATIONSOURCE and REPLICATIONDEST privileges

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142152)
<!-- Reviewable:end -->

msbutler committed:
jobsauth: remove getLegacyPayload arg from jobsauth.Authorize

A new jobsauth.AuthorizeAllowLegacyAuth func is added for callsites that still
we require legacy jobs auth.

Epic: none

Release note: none

msbutler committed:
crosscluster/producer: use job auth with provided stream id for some builtins

Previously, all replication builtins required authorization via the REPLICATION
privilege. With this patch, all builtins with a stream id arg can authorize if
the user connecting to the source cluser has control privs on the producer job.

In other words, once a user has created the producer job associated with the
LDR or PCR stream, authorized via their REPLICATION priv, they can run
subsequent replication commands, authorized via job auth.

This patch paves the way for subsequent privilege refactors, like table level
LDR authentication, without the need to refactor the replication builtins args.

Epic: none

Release note: none

msbutler committed:
add ranomized non root user testing

msbutler committed:
crosscluster/producer: auth replication creation with REPLICATIONSOURCE priv

Release note (ops change): previously, the user supplied in the source cluster
connection in PCR and LDR required the REPLICATION priv, which will be
deprecated in 25.2 (undocumented, temporarily supported). Instead, the user
will need the new REPLICATION SOURCE priv.

msbutler committed:
crosscluster: user that creates LDR or PCR stream requires REPLICATIONDEST priv

msbutler committed:
crosscluster/producer: enable table level replication source privs
